tools/ion-bench/src/ion/writing.h

```diff
@@ -0,0 +1,237 @@
+
+      iERR write(const std::optional<bool> &b) {
+         if (b) {
+            return ion_writer_write_bool(_writer, *b);
+         } else {
+            return ion_writer_write_typed_null(_writer, tid_BOOL);
+         }
+      }
+
+      iERR write(const std::optional<double> &d) {
+         if (d) {
+            return ion_writer_write_double(_writer, *d);
+         } else {
+            return ion_writer_write_typed_null(_writer, tid_FLOAT);
+         }
+      }
+
+      iERR write(const std::optional<IonDecimalPtr> &d) {
+         if (d.has_value()) {
+            return ion_writer_write_ion_decimal(_writer, (*d).get());
+         } else {
+            return ion_writer_write_typed_null(_writer, tid_DECIMAL);
+         }
+      }
+
+      iERR write(const std::optional<Symbol> &sym) {
+         if (sym.has_value()) {
+            ION_STRING str = {0};
+            ion_string_assign_cstr(&str, (char *)sym->value.c_str(), sym->value.length());
+            return ion_writer_write_symbol(_writer, &str);
+         } else {
+            return ion_writer_write_typed_null(_writer, tid_SYMBOL);
+         }
+      }
+
+      iERR write(const std::optional<std::shared_ptr<ION_TIMESTAMP>> &ts) {
+         if (ts.has_value()) {
+            return ion_writer_write_timestamp(_writer, ts.value().get());
+         } else {
+            return ion_writer_write_typed_null(_writer, tid_TIMESTAMP);
+         }
+      }
+};
+
+#define WRITE_ION_TYPE(writer, tid, tpe, val)                                                      \
+   case tid: {                                                                                     \
+      std::optional<tpe> t = val.value ?                                                           \
+                std::make_optional(std::any_cast<tpe>(val.value.value()))                          \
+              : std::nullopt;                                                                      \
+      err = writer.write(t);                                                                       \
+      if (err != IERR_OK) { printf("ERROR writing " #tpe ": %d\n", err); }                         \
+      break;                                                                                       \
+   }
+
+template <Format F>
+iERR write(BufferWriter<F> &writer, const IonData &val) {
+   iERR err = IERR_NOT_IMPL;
+   writer.write_annotations(val.annotations);
+   switch (val.tpe) {
+      WRITE_ION_TYPE(writer, tid_STRING_INT, std::string, val);
+      WRITE_ION_TYPE(writer, tid_INT_INT, int64_t, val);
+      WRITE_ION_TYPE(writer, tid_FLOAT_INT, double, val);
+      WRITE_ION_TYPE(writer, tid_DECIMAL_INT, IonDecimalPtr, val);
+      WRITE_ION_TYPE(writer, tid_BOOL_INT, bool, val);
+      WRITE_ION_TYPE(writer, tid_SYMBOL_INT, Symbol, val);
+      WRITE_ION_TYPE(writer, tid_TIMESTAMP_INT, std::shared_ptr<ION_TIMESTAMP>, val);
+      default:
+         printf("Attempt to write unknown type: %d\n", val.tpe);
+         break;
```

Comment:
Are we missing support for `null.null`, blob, and clob?

Reply:
Agh, yea. I'll include those in the follow-up PR.